File: chrome/android/java/src/org/chromium/chrome/browser/tab/TabSessionIsolationBridge.java
// Copyright 2024 Kiwi Browser Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license.

package org.chromium.chrome.browser.tab;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.profiles.Profile;

/**
 * JNI bridge for native TabSessionIsolationManager.
 *
 * <p>This class exposes the C++ TabSessionIsolationManager to Java.
 * All methods must be called on the UI thread.
 */
@JNINamespace("tab_session_isolation")
public class TabSessionIsolationBridge {

    /**
     * Creates a new isolated session.
     *
     * @param profile The current user profile.
     * @return A UUID string (isolation_key) identifying the session,
     *         or an empty string if the maximum number of sessions
     *         ({@code TabSessionIsolationManager::kMaxIsolatedSessions}) is
     *         already reached.
     */
    public static String createIsolationSession(Profile profile) {
        return TabSessionIsolationBridgeJni.get().createIsolationSession(profile);
    }

    /**
     * Destroys an isolated session and schedules asynchronous deletion of all
     * on-disk storage associated with it (cookies, IndexedDB, localStorage,
     * cache, service workers).
     *
     * <p>Must be called when the corresponding tab is closed.
     *
     * @param profile       The current user profile.
     * @param isolationKey  The UUID returned by {@link #createIsolationSession}.
     */
    public static void destroyIsolationSession(Profile profile, String isolationKey) {
        if (isolationKey == null || isolationKey.isEmpty()) return;
        TabSessionIsolationBridgeJni.get().destroyIsolationSession(profile, isolationKey);
    }

    /**
     * Returns the number of currently active isolated sessions.
     *
     * @return Active session count (0 to kMaxIsolatedSessions).
     */
    public static int getActiveSessionCount() {
        return TabSessionIsolationBridgeJni.get().getActiveSessionCount();
    }

    /**
     * Returns {@code true} if a new isolated session can be created.
     *
     * @return {@code false} if kMaxIsolatedSessions is already reached.
     */
    public static boolean canCreateIsolatedSession() {
        return TabSessionIsolationBridgeJni.get().canCreateIsolatedSession();
    }

    @NativeMethods
    interface Natives {
        String createIsolationSession(Profile profile);
        void destroyIsolationSession(Profile profile, String isolationKey);
        int getActiveSessionCount();
        boolean canCreateIsolatedSession();
    }

    private TabSessionIsolationBridge() {}
}

File: chrome/browser/tab_session_isolation/android/tab_session_isolation_bridge.cc
// Copyright 2024 Kiwi Browser Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license.

#include "chrome/browser/tab_session_isolation/android/tab_session_isolation_bridge.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "base/android/scoped_java_ref.h"
#include "chrome/android/chrome_jni_headers/TabSessionIsolationBridge_jni.h"
#include "chrome/browser/profiles/profile.h"
#include "chrome/browser/profiles/profile_android.h"
#include "chrome/browser/tab_session_isolation/tab_session_isolation_manager.h"
#include "content/public/browser/browser_context.h"

using base::android::ConvertJavaStringToUTF8;
using base::android::ConvertUTF8ToJavaString;
using base::android::JavaParamRef;
using base::android::ScopedJavaLocalRef;

namespace tab_session_isolation {

// ---------------------------------------------------------------------------
// JNI exported functions – called from Java via generated JNI glue.
// Each function is named JNI_<ClassName>_<MethodName> per Chromium convention.
// ---------------------------------------------------------------------------

// Creates a new isolated session.
// Returns the isolation_key (UUID) or an empty string when the max is reached.
static ScopedJavaLocalRef<jstring>
JNI_TabSessionIsolationBridge_CreateIsolationSession(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(env, j_profile);
  if (!profile) {
    return ConvertUTF8ToJavaString(env, "");
  }

  TabSessionIsolationManager* manager =
      TabSessionIsolationManager::GetInstance();

  std::string isolation_key =
      manager->CreateIsolationSession(profile->GetOriginalProfile());

  return ConvertUTF8ToJavaString(env, isolation_key);
}

// Destroys an isolated session and schedules deletion of its on-disk storage.
static void JNI_TabSessionIsolationBridge_DestroyIsolationSession(
    JNIEnv* env,
    const JavaParamRef<jobject>& j_profile,
    const JavaParamRef<jstring>& j_isolation_key) {
  Profile* profile = ProfileAndroid::FromProfileAndroid(env, j_profile);
  if (!profile) {
    return;
  }

  std::string isolation_key = ConvertJavaStringToUTF8(env, j_isolation_key);
  if (isolation_key.empty()) {
    return;
  }

  TabSessionIsolationManager::GetInstance()->DestroyIsolationSession(
      profile->GetOriginalProfile(), isolation_key);
}

// Returns the number of currently active isolated sessions.
static jint JNI_TabSessionIsolationBridge_GetActiveSessionCount(JNIEnv* env) {
  return TabSessionIsolationManager::GetInstance()->GetActiveSessionCount();
}

// Returns true if a new isolated session can be created.
static jboolean JNI_TabSessionIsolationBridge_CanCreateIsolatedSession(
    JNIEnv* env) {
  return TabSessionIsolationManager::GetInstance()->CanCreateIsolatedSession()
             ? JNI_TRUE
             : JNI_FALSE;
}

}  // namespace tab_session_isolation
